File: eShopModernizedJava/src/main/java/com/eshop/catalog/controller/PicController.java
package com.eshop.catalog.controller;

import com.eshop.catalog.service.CatalogService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.core.io.ClassPathResource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;

import java.io.IOException;

@Controller
public class PicController {

    private static final Logger log = LoggerFactory.getLogger(PicController.class);

    private final CatalogService catalogService;

    public PicController(CatalogService catalogService) {
        this.catalogService = catalogService;
    }

    @GetMapping("/items/{catalogItemId}/pic")
    public ResponseEntity<byte[]> getImage(@PathVariable int catalogItemId) {
        log.info("Now loading... /items/{}/pic", catalogItemId);

        if (catalogItemId <= 0) {
            return ResponseEntity.badRequest().build();
        }

        var item = catalogService.findCatalogItem(catalogItemId);
        if (item == null) {
            return ResponseEntity.notFound().build();
        }

        try {
            String fileName = item.getPictureFileName();
            if (fileName.contains("..") || fileName.contains("/") || fileName.contains("\\")) {
                return ResponseEntity.badRequest().build();
            }
            ClassPathResource resource = new ClassPathResource("static/pics/" + fileName);
            if (!resource.exists()) {
                return ResponseEntity.notFound().build();
            }

            byte[] imageBytes = resource.getContentAsByteArray();
            String extension = getFileExtension(item.getPictureFileName());
            String mimeType = getImageMimeType(extension);

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.parseMediaType(mimeType));
            return new ResponseEntity<>(imageBytes, headers, HttpStatus.OK);
        } catch (IOException e) {
            log.error("Error reading image file for item {}", catalogItemId, e);
            return ResponseEntity.internalServerError().build();
        }
    }

    private String getFileExtension(String filename) {
        int dotIndex = filename.lastIndexOf('.');
        if (dotIndex < 0) {
            return "";
        }
        return filename.substring(dotIndex).toLowerCase();
    }

    private String getImageMimeType(String extension) {
        return switch (extension) {
            case ".png" -> "image/png";
            case ".gif" -> "image/gif";
            case ".jpg", ".jpeg" -> "image/jpeg";
            case ".bmp" -> "image/bmp";
            case ".tiff" -> "image/tiff";
            case ".wmf" -> "image/wmf";
            case ".jp2" -> "image/jp2";
            case ".svg" -> "image/svg+xml";
            default -> "application/octet-stream";
        };
    }
}

File: eShopModernized/src/eShop.Catalog.Api/Controllers/PicController.cs
using eShop.Catalog.Domain.Abstractions;
using eShop.Shared.Configuration;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Options;

namespace eShop.Catalog.Api.Controllers;

/// <summary>
/// Catalog item picture endpoint, ported from the legacy MVC
/// <c>eShopLegacyMVC.Controllers.PicController</c>. Pictures are read from
/// <see cref="CatalogOptions.PicsFolder"/> (relative to the content root) instead of
/// <c>Server.MapPath("~/Pics")</c>.
/// </summary>
[ApiController]
public class PicController : ControllerBase
{
    /// <summary>Route name of the picture endpoint, kept from the legacy controller.</summary>
    public const string GetPicRouteName = "GetPicRouteTemplate";

    private readonly ICatalogService _catalogService;
    private readonly IWebHostEnvironment _environment;
    private readonly CatalogOptions _options;
    private readonly ILogger<PicController> _logger;

    /// <summary>Creates the controller.</summary>
    public PicController(
        ICatalogService catalogService,
        IWebHostEnvironment environment,
        IOptions<CatalogOptions> options,
        ILogger<PicController> logger)
    {
        ArgumentNullException.ThrowIfNull(options);

        _catalogService = catalogService;
        _environment = environment;
        _options = options.Value;
        _logger = logger;
    }

    /// <summary>
    /// Returns the picture of a catalog item: 400 for a non-positive identifier, 404 when the item
    /// or its picture file does not exist, otherwise the file with its image content type.
    /// </summary>
    [HttpGet("items/{catalogItemId:int}/pic", Name = GetPicRouteName)]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetPic(int catalogItemId, CancellationToken cancellationToken)
    {
        _logger.LogInformation("Now loading picture of catalog item {CatalogItemId}", catalogItemId);

        if (catalogItemId <= 0)
        {
            return BadRequest();
        }

        var item = await _catalogService.FindCatalogItemAsync(catalogItemId, cancellationToken);

        if (item is null)
        {
            return NotFound();
        }

        var path = Path.Combine(PicsFolderPath(), item.PictureFileName);

        if (!System.IO.File.Exists(path))
        {
            _logger.LogWarning(
                "Picture file {PictureFileName} of catalog item {CatalogItemId} was not found",
                item.PictureFileName,
                catalogItemId);

            return NotFound();
        }

        var buffer = await System.IO.File.ReadAllBytesAsync(path, cancellationToken);

        return File(buffer, GetImageMimeTypeFromImageFileExtension(Path.GetExtension(item.PictureFileName)));
    }

    private string PicsFolderPath() =>
        Path.IsPathRooted(_options.PicsFolder)
            ? _options.PicsFolder
            : Path.Combine(_environment.ContentRootPath, _options.PicsFolder);

    /// <summary>Extension to content type mapping, kept identical to the legacy controller.</summary>
    private static string GetImageMimeTypeFromImageFileExtension(string extension) => extension switch
    {
        ".png" => "image/png",
        ".gif" => "image/gif",
        ".jpg" or ".jpeg" => "image/jpeg",
        ".bmp" => "image/bmp",
        ".tiff" => "image/tiff",
        ".wmf" => "image/wmf",
        ".jp2" => "image/jp2",
        ".svg" => "image/svg+xml",
        _ => "application/octet-stream",
    };
}
